src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/DisposeAnalysis/DisposeAnalysis.DisposeDataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,302 @@
+                return value;
+            }
+
+            public override DisposeAbstractValue VisitArgument(IArgumentOperation operation, object argument)
+            {
+                var value = base.VisitArgument(operation, argument);
+
+                // Discover if a disposable object is being passed into the creation method for this new disposable object
+                // and if the new dispoable object assumes ownership of that passed in disposable object.
+                if (operation.Parent is IObjectCreationOperation objectCreation &&
+                    objectCreation.Arguments.Length == 1 &&
+                    objectCreation.Constructor.Parameters.Length == 1 &&
+                    _disposeOwnershipTransferLikelyTypes.Contains(objectCreation.Constructor.Parameters[0].Type))
+                {
+                    HandlePossibleEscapingOperation(operation, operation.Value);
+                }
+                else if (operation.Parameter.RefKind != RefKind.None)
+                {
+                    // Argument passed by ref/out is considered escaped.
+                    HandlePossibleEscapingOperation(operation, operation.Value);
+                }
+
+                return value;
+            }
+
+            public override DisposeAbstractValue VisitReturn(IReturnOperation operation, object argument)
+            {
+                var value = base.VisitReturn(operation, argument);
+                HandlePossibleEscapingOperation(operation, operation.ReturnedValue);
+                return value;
+            }
+
+            public override DisposeAbstractValue VisitUsing(IUsingOperation operation, object argument)
+            {
+                var value = base.VisitUsing(operation, argument);
+                if (operation.Resources is IVariableDeclarationGroupOperation varDeclGroup)
+                {
+                    var variableDeclarationInitializers = varDeclGroup.Declarations.Select(declaration => declaration.Initializer?.Value).WhereNotNull();
+                    var variableDeclaratorInitializers = varDeclGroup.Declarations.SelectMany(declaration => declaration.Declarators).Select(declarator => declarator.Initializer?.Value).WhereNotNull();
+                    foreach (var disposedInstance in variableDeclarationInitializers.Concat(variableDeclaratorInitializers))
+                    {
+                        HandleDisposingOperation(operation, disposedInstance);
+                    }
+                }
+                else if (operation.Resources != null)
+                {
+                    HandleDisposingOperation(operation, operation.Resources);
+                }
+
+                return value;
+            }
+
+            public override DisposeAbstractValue VisitConversion(IConversionOperation operation, object argument)
+            {
+                var value = base.VisitConversion(operation, argument);
```

Comment:
Should we do anything if the thing being converted is disposable, but is being converted to a non-disposable type?

Reply:
We just retain the current state for the disposable location for the converted object. Subsequently, user might cast it back to the disposable object and invoke dispose, at which point we will change the dispose state appropriately. I will add a unit test for this scenario.